helix-core/src/main/java/org/apache/helix/controller/dataproviders/BaseControllerDataProvider.java

```diff
@@ -241,11 +244,85 @@ private void refreshClusterConfig(final HelixDataAccessor accessor,
+      if (needReformatBatchedDIsabledInstance(_clusterConfig) && updateBatchDisableFormat(
+          accessor)) {
+        // read from zkz one more time
+        LogUtil.logInfo(logger, getClusterEventId(), String
+            .format("Clean ClusterConfig change for cluster %s, pipeline %s", _clusterName,
+                getPipelineName()));
+        _clusterConfig = accessor.getProperty(accessor.keyBuilder().clusterConfig());
+      }
       refreshAbnormalStateResolverMap(_clusterConfig);
     } else {
-      LogUtil.logInfo(logger, getClusterEventId(), String.format(
-          "No ClusterConfig change for cluster %s, pipeline %s", _clusterName, getPipelineName()));
+      LogUtil.logDebug(logger, getClusterEventId(), String
+          .format("No ClusterConfig change for cluster %s, pipeline %s", _clusterName,
+              getPipelineName()));
+    }
+  }
+
+  // TODO: This function is used to clean up incompatible batched disabled instances format for
+  // "DISABLED_INSTANCES" introduced in 1.0.3.0. This temp change shoul dbe reverted after 1.0.5.0 \
+  // or later version.
+  private boolean updateBatchDisableFormat(final HelixDataAccessor accessor) {
+    return accessor
+        .updateProperty(accessor.keyBuilder().clusterConfig(), new DataUpdater<ZNRecord>() {
+          @Override
+          public ZNRecord update(ZNRecord currentData) {
+            if (currentData == null) {
+              throw new HelixException(
+                  "Cluster: " + _clusterConfig.getClusterName() + ": cluster config is null");
+            }
+
+            ClusterConfig clusterConfig = new ClusterConfig(currentData);
+            Map<String, String> disabledInstances = clusterConfig.getDisabledInstances();
+            Map<String, String> disabledInstancesWithInfo =
+                clusterConfig.getDisabledInstancesWithInfo();
+
+            ClusterConfig newClusterConfig = new ClusterConfig(currentData);
+            Map<String, String> newDisabledInstances =
+                new TreeMap<>(newClusterConfig.getDisabledInstances());
+            Map<String, String> newDisabledInstancesWithInfo =
+                new TreeMap<>(newClusterConfig.getDisabledInstancesWithInfo());
+
+            // iterate through all k,v pairs and check for string format.
+            for (Map.Entry<String, String> instanceInfo : disabledInstances.entrySet()) {
+              String instanceName = instanceInfo.getKey();
+              if (!StringUtils.isNumeric(instanceInfo.getValue())) {
+                newDisabledInstances.put(instanceName,
+                    ConfigStringUtil.parseConcatenatedConfig(instanceInfo.getValue()).get(
+                        ClusterConfig.ClusterConfigProperty.HELIX_ENABLED_DISABLE_TIMESTAMP
+                            .name()));
+                newDisabledInstancesWithInfo.put(instanceName, instanceInfo.getValue());
+              } else {
+                if (!disabledInstancesWithInfo.containsKey(instanceName)) {
```

Comment:
Nit: can these two lines be reformatted as `} else if () {`?

Reply:
Updated.